x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java

```diff
@@ -64,427 +114,1074 @@
+        } else {
+            return rawJavaTimestampFormat;
+        }
+    }
 
-        return (requiredFormat == null) ? null :
-            FRACTIONAL_SECOND_TIMESTAMP_FORMAT_PATTERN.matcher(requiredFormat).replaceFirst(DEFAULT_FRACTIONAL_SECOND_FORMAT);
+    /**
+     * These are still used by Logstash.
+     * @return A list of Joda timestamp formats that correspond to the detected Java timestamp formats.
+     */
+    public List<String> getJodaTimestampFormats() {
+        List<String> javaTimestampFormats = getJavaTimestampFormats();
+        return (javaTimestampFormats == null) ? null : javaTimestampFormats.stream()
+            .map(format -> format.replace("yy", "YY").replace("XXX", "ZZ").replace("XX", "Z")).collect(Collectors.toList());
     }
 
-    private static TimestampMatch makeTimestampMatch(CandidateTimestampFormat chosenTimestampFormat, int chosenIndex,
-                                                     String preface, String matchedDate, String epilogue) {
-        Tuple<Character, Integer> fractionalSecondsInterpretation = interpretFractionalSeconds(matchedDate);
-        List<String> jodaTimestampFormats = chosenTimestampFormat.jodaTimestampFormats;
-        List<String> javaTimestampFormats = chosenTimestampFormat.javaTimestampFormats;
-        Pattern simplePattern = chosenTimestampFormat.simplePattern;
-        char separator = fractionalSecondsInterpretation.v1();
-        if (separator != DEFAULT_FRACTIONAL_SECOND_SEPARATOR) {
-            jodaTimestampFormats = jodaTimestampFormats.stream()
-                .map(jodaTimestampFormat -> jodaTimestampFormat.replace(DEFAULT_FRACTIONAL_SECOND_SEPARATOR, separator))
-                .collect(Collectors.toList());
-            javaTimestampFormats = javaTimestampFormats.stream()
-                .map(javaTimestampFormat -> javaTimestampFormat.replace(DEFAULT_FRACTIONAL_SECOND_SEPARATOR, separator))
-                .collect(Collectors.toList());
-            if (jodaTimestampFormats.stream().noneMatch(jodaTimestampFormat -> jodaTimestampFormat.startsWith("UNIX"))) {
-                String patternStr = simplePattern.pattern();
-                int separatorPos = patternStr.lastIndexOf(DEFAULT_FRACTIONAL_SECOND_SEPARATOR);
-                if (separatorPos >= 0) {
-                    StringBuilder newPatternStr = new StringBuilder(patternStr);
-                    newPatternStr.replace(separatorPos, separatorPos + 1, ((separator == '.') ? "\\" : "") + separator);
-                    simplePattern = Pattern.compile(newPatternStr.toString());
+    /**
+     * Does the parsing the timestamp produce different results depending on the timezone of the parser?
+     * I.e., does the textual representation NOT define the timezone?
+     */
+    public boolean hasTimezoneDependentParsing() {
+        if (matchedFormats.isEmpty()) {
+            if (errorOnNoTimestamp) {
+                throw new IllegalStateException("No matched timestamp formats");
```

Comment:
We should include the `why` in this exception.

Reply:
None of these `IllegalStateException`s in the getters are user facing error messages.  They mean the programmer has made a mistake by asking for results from an object they didn't add any samples to (because if they added a sample when `errorOnNoTimestamp` is `true` then _that_ would have thrown an exception).

It's obviously confusing as it is now.  I think I should change these exceptions to `assert`s instead since they're detecting usage errors by the programmer.  The return value for when assertions are disabled can just be the `null`/empty/false/0 that gets returned when `errorOnNoTimestamp` is `false`.  If you think that sounds OK then I'll change it tomorrow, but let me know if you think of a better approach.